librz/core/linux_heap_glibc.c

```diff
@@ -1406,84 +1840,217 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
 			}
 		}
 
-		switch (format_out) {
-		case 'c':
-			GH(print_heap_chunk_simple)
-			(core, prev_chunk_addr, status);
-			rz_cons_newline();
-			break;
-		case 'v':
+		RzHeapChunkListItem *block = malloc(sizeof(RzHeapChunkListItem));
+		block->addr = prev_chunk_addr;
+		block->status = status;
+		block->size = prev_chunk_size;
+		rz_list_append(chunks, block);
+	}
+	free(cnk);
+	free(cnk_next);
+	return chunks;
+}
+
+RZ_IPI RzCmdStatus GH(rz_cmd_arena_print_handler)(RzCore *core, int argc, const char **argv) {
+	static GHT m_arena = GHT_MAX;
+	RzConsPrintablePalette *pal = &rz_cons_singleton()->context->pal;
+	MallocState *main_arena = RZ_NEW0(MallocState);
+	if (!main_arena) {
+		return RZ_CMD_STATUS_ERROR;
+	}
+	if (!GH(rz_resolve_main_arena)(core, &m_arena)) {
+		free(main_arena);
+		return RZ_CMD_STATUS_ERROR;
+	}
+	if (!GH(rz_update_main_arena)(core, m_arena, main_arena)) {
+		free(main_arena);
+		return RZ_CMD_STATUS_ERROR;
+	}
+	RzList *arenas_list = GH(rz_get_arenas_list)(core, m_arena, main_arena);
+	RzListIter *iter, *iter2;
+	MallocState *pos;
+	ut64 addr;
+	rz_list_foreach (arenas_list, iter2, pos) {
+		addr = pos->GH(next);
+	}
+	bool flag = 0;
+	rz_list_foreach (arenas_list, iter, pos) {
+		if (!flag) {
+			flag = true;
+			rz_cons_printf("Main arena  (addr=");
+
+		} else {
+			rz_cons_printf("Thread arena(addr=");
+		}
+		PRINTF_YA("0x%" PFMT64x, (ut64)addr);
+		rz_cons_printf(", lastRemainder=");
+		PRINTF_YA("0x%" PFMT64x, (ut64)pos->GH(last_remainder));
+		rz_cons_printf(", top=");
+		PRINTF_YA("0x%" PFMT64x, (ut64)pos->GH(top));
+		rz_cons_printf(", next=");
+		PRINTF_YA("0x%" PFMT64x, (ut64)pos->GH(next));
+		if (pos->attached_threads) {
+			rz_cons_printf(")\n");
+		} else {
+			rz_cons_printf(", free)\n");
+		}
+		addr = (ut64)pos->GH(next);
+	}
+	rz_list_free(arenas_list);
+	free(main_arena);
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
```

Comment:
I'd have this function with `GHT m_state` argument instead of argc/argv. That woulld make calling them "easier" for other parts of the code.